### PR TITLE
feat: hrid redesign, designer settings, template editor and TemplatedStringField changes

### PR DESCRIPTION
**Pull request by PeterBaker0**
# feat: hrid redesign, designer settings, template editor and TemplatedStringField changes

## JIRA Ticket

[BSS-398](https://jira.csiro.au/browse/BSS-398)

## Description

This PR addresses a number of limitations in the existing HRID implementation. Changes included

- implementation of new 'Form settings' module at the top of designer for forms which includes summary fields, layout settings and,
- new `hridField` selection at Form level which allows specific selection of any required text type field as a HRID
- removal of hrid prefix system from designer, and migration for old notebooks to adopt the new style (app is backwards compatible with old hrid prefix fields)
- updates to HRID methods in the app to look for the new HRID first, then fall back to previous logic or `record_id` as the last resort
- upgrades to the TemplatedStringField capability including
  - system context variables including created by and created at which can be injected in the editor and are populated when available at run time
  - (designer) whole new interactive template editor which includes variables, text, if, and unless blocks
  - (designer) template preview capability allowing you to test template resolution with different variable values
- addition of `hidden` prop which allows you to hide a field from display (this is equivalent to a conditionFn excluding it) @stevecassidy this may have implications for export but we can refine this - addition of a panel for designer to toggle this, only enabled on templated string field currently

## How to Test

Test designer by

- creating a text field
- creating a TemplatedStringField which references this field and/or system variables 
- do this manually and using the visual editor - test out the preview
- select this field as the HRID
- test out this works in the app

I'd also check the summary fields - make sure these work + layouts. 

Probably worth checking backwards compatibility with old HRID, and that the migration works - I've tried to test both and seem fine.

## Additional Information

Some corners were cut here, mostly due to the poor implementation quality of the HRID logic and general form record handling. 

i.e. 

- in many places, the fact that hrid was the prefix for the name of the field was heavily depended on in that context to determine which field to use (so I've had to reverse engineer from the ui_spec where available based on the uniqueness of field names between forms)
- the HRID is depended on/used too much in general across the app - there is no need for the HRID to be embedded into the data and depended on, it should be rendered when needed - it is just a transient 'view' over form data which should only be necessary for visualising the data or maybe during export

## Checklist

- [x] I have confirmed all commits have been signed.
- [x] I have added JSDoc style comments to any new functions or classes.
- [x] Relevant documentation such as READMEs, guides, and class comments are updated.

### CODE REVIEW

**Review by PeterBaker0 — COMMENTED** [review]
Accidentally reviewed

File: app/package.json [hunk 1]
```diff
@@ -98,6 +98,7 @@
         "@worker-tools/json-stream": "^0.1.0-pre.12",
         "animate.css": "^4.1.1",
         "assert": "^2.0.0",
+        "base64-arraybuffer": "^1.0.2",
```
[the remark this answers is not in the document]
Comment (stevecassidy) [reply app/package.json:101]: Consider using Buffer instead (https://www.webdevtutor.net/blog/typescript-atob-deprecated).

File: designer/src/components/TemplateBuilder.tsx [hunk 2]
```diff
@@ -0,0 +1,763 @@
...
+ * Converts a Mustache AST token into our TemplateBlock format
+ */
+const convertMustacheTokenToBlock = (token: any): TemplateBlock | null => {
+  const [tokenType, content, ...rest] = token;
+
+  try {
+    switch (tokenType) {
+      case 'text':
+        return {
+          id: Math.random().toString(36).substring(2, 9),
```
Comment (stevecassidy) [inline designer/src/components/TemplateBuilder.tsx:116]: This is not guaranteed to be unique, what happens when you get a collision? Since you use this code a lot it should at least be a fn. Suggest just using an incrementing variable to get a unique id.

File: designer/src/components/TemplateBuilder.tsx [hunk 3]
```diff
@@ -0,0 +1,763 @@
...
+          <Grid item xs={12} sm>
+            <FormControl fullWidth>
+              <InputLabel>Variable</InputLabel>
+              <Select
+                value={block.content}
+                label="Variable"
+                onChange={e => updateBlockContent(e.target.value)}
+              >
+                <MenuItem value="">Choose variable...</MenuItem>
+                {allVariables.map(v => (
```
Comment (stevecassidy) [inline designer/src/components/TemplateBuilder.tsx:235]: You should only be able to add fields from this form to the expression. Doesn't make sense to add a field from another form since it can't have a value here.

File: designer/src/components/TemplateBuilder.tsx [hunk 4]
```diff
@@ -0,0 +1,763 @@
...
+            <Select
+              value={block.type}
+              label="Block Type"
+              onChange={e =>
+                onUpdate({...block, type: e.target.value as BlockType})
+              }
+            >
+              <MenuItem value="text">Text</MenuItem>
+              <MenuItem value="variable">Variable</MenuItem>
+              <MenuItem value="if">Conditional</MenuItem>
```
Comment (stevecassidy) [inline designer/src/components/TemplateBuilder.tsx:219]: Conditional is not clear. When I make a conditional block I choose a variable but then I'm done - i then need to add another block that gets indented... doesn't seem very clear to me.  Maybe just some helper text will do?

Comment (stevecassidy) [reply designer/src/components/TemplateBuilder.tsx:219]: Also 'Unless' is also a conditional, so maybe 'Conditional' should be 'If'?

File: designer/src/components/Fields/HiddenToggle.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  FormControlLabel,
+  Paper,
+  Stack,
+  Tooltip,
+  Grid,
+} from '@mui/material';
+import {useAppDispatch, useAppSelector} from '../../state/hooks';
+import {FieldType} from '../../state/initial';
+
+/**
```
Comment (stevecassidy) [inline designer/src/components/Fields/HiddenToggle.tsx:15]: Ryan's PR also adds a hidden toggle, so duplicating work here.

File: designer/src/fields.tsx [hunk 6]
```diff
@@ -402,16 +402,15 @@ const fields: {[key: string]: FieldType} = {
...
-      label: 'Human Readable ID',
+      label: 'Templated String Field',
       fullWidth: true,
-      name: 'hrid-field',
-      helperText: 'Human Readable ID',
+      name: 'templated-field',
+      helperText: 'Templated String Field',
       required: true,
       template: ' {{}}',
       InputProps: {
```
Comment (stevecassidy) [inline designer/src/fields.tsx:411]: InputProps can probably go

File: app/src/gui/components/record/branchingLogic.tsx [hunk 7]
```diff
@@ -44,7 +44,13 @@ export function getFieldsMatchingCondition(
   if (modified.length > 0 || fieldNames.length === 0) {
     // filter the whole set of views
     const result = allFields.filter(field => {
-      return ui_specification.fields[field].conditionFn(values);
+      const fieldDetails = ui_specification.fields[field];
+      return (
+        // Visibility condition function
+        fieldDetails.conditionFn(values) &&
+        // Hidden explicitly in element props - e.g. templated field
+        !fieldDetails['component-parameters']?.ElementProps?.hidden
```
Comment (stevecassidy) [inline app/src/gui/components/record/branchingLogic.tsx:52]: Do we really want this to be in ElementProps? I'd say put it in component-parameters directly. I want to get rid of ElementProps and it's friends.

Comment (stevecassidy) [reply app/src/gui/components/record/branchingLogic.tsx:52]: Ryan's PR puts this in component-parameters

File: app/src/gui/fields/TemplatedStringField.tsx [hunk 8]
```diff
@@ -53,52 +44,14 @@ export class TemplatedStringField extends React.Component<
...
-      if (value !== '')
-        if (this.props.form.errors[this.props.field.name] !== undefined)
-          this.props.form.setFieldError(this.props.field.name, undefined);
-    }
-  }
-
   render() {
     return (
       <>
         <MuiTextField
```
Comment (stevecassidy) [inline app/src/gui/fields/TemplatedStringField.tsx:50]: Do you want to render this as something other than a text field? We have a task to make it not look like it can be edited. Maybe just a plain text render would be better?

File: app/src/gui/components/record/relationships/RelatedInformation.tsx [hunk 9]
```diff
@@ -674,7 +715,25 @@ export async function getParentPersistenceData(
         )
       );
 
-      const parent_hrid = getHRIDValue(record_id, latest_record?.data);
+      // Here we reverse engineer the correct hrid field to look for since we
```
Comment (stevecassidy) [inline app/src/gui/components/record/relationships/RelatedInformation.tsx:718]: Repeated code from line 576 - factor out?

File: app/src/gui/components/record/relationships/RelatedInformation.tsx [hunk 10]
```diff
@@ -739,7 +819,23 @@ export async function getDetailRelatedInformation(
     values
   );
 
-  const hrid = getHRIDValue(record_id, values);
+  // Here we reverse engineer the correct hrid field to look for since we
```
Comment (stevecassidy) [inline app/src/gui/components/record/relationships/RelatedInformation.tsx:822]: And again, definitely factor out.

File: app/src/gui/components/record/relationships/RelatedInformation.tsx [hunk 11]
```diff
@@ -897,14 +990,17 @@ export async function addRecordLink(
...
+    // here we are trusting that Record has enough of the fields of
+    // RecordMetadata for the purposes of the caller until such time as we
+    // rationalise the Record types
     child_record_meta = latest_record as unknown as RecordMetadata;
     // it's missing the HRID so grab it here
-    child_record_meta.hrid = getHRIDValue(
-      child_record_meta.record_id,
-      child_record_meta.data
-    );
+    // Just use record id here and see what happens ¯\_(ツ)_/¯
```
Comment (stevecassidy) [inline app/src/gui/components/record/relationships/RelatedInformation.tsx:998]: What happens!? I'm dying to know!

Comment (stevecassidy) [reply app/src/gui/components/record/relationships/RelatedInformation.tsx:998]: Child links seem to be fine.

Comment (stevecassidy) [reply app/src/gui/components/record/relationships/RelatedInformation.tsx:998]: Ah no, so this populates the list of records if you select 'add link to an existing XXX', we now see a list of record IDs rather than the HRIDs. So we really want to have the HRID here.

File: app/src/gui/fields/TakePhoto.tsx [hunk 12]
```diff
@@ -28,6 +28,7 @@ import {Alert, Box, Link, Paper, Typography, useTheme} from '@mui/material';
 import IconButton from '@mui/material/IconButton';
```
Comment (stevecassidy) [inline app/src/gui/fields/TakePhoto.tsx:28]: Why are you fixing TakePhoto in this PR?

File: app/src/gui/fields/TakePhoto.tsx [hunk 13]
```diff
@@ -28,6 +28,7 @@ import {Alert, Box, Link, Paper, Typography, useTheme} from '@mui/material';
...
 import ImageListItem from '@mui/material/ImageListItem';
 import ImageListItemBar from '@mui/material/ImageListItemBar';
+import {decode} from 'base64-arraybuffer';
```
Comment (stevecassidy) [inline app/src/gui/fields/TakePhoto.tsx:31]: See comment about using Buffer.

File: app/src/logging.tsx [hunk 14]
```diff
@@ -125,6 +125,7 @@ export const ErrorPage = () => {
 
 export const logError = (error: any) => {
   if (BUGSNAG_KEY) {
+    console.error(error);
```
Comment (stevecassidy) [inline app/src/logging.tsx:128]: we don't want error logging in production please!

File: app/src/utils/formUtilities.ts [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+
+/**
+ * Formats a timestamp into a date-time string in the format "DD/MM/YY H:MMam/pm"
+ *
+ * @param timestamp - Unix timestamp in milliseconds (e.g., from Date.now())
+ * @returns Formatted date-time string or empty string if input is invalid
+ *
+ * @throws Never - Returns empty string for all error cases
+ *
+ * Handles the following edge cases:
```
Comment (stevecassidy) [inline app/src/utils/formUtilities.ts:34]: Tests?

File: app/src/utils/formUtilities.ts [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+/**
+ * Converts the RecordContext into an object mapping from key -> value for use
+ * in template replacement
+ * @param context
+ * @returns The context values to inject
+ */
+function contextToTemplate(context: RecordContext): ValuesObject {
+  const vals: ValuesObject = {};
+  if (context.createdBy) {
+    vals[CREATOR_NAME_ID] = context.createdBy;
```
Comment (stevecassidy) [inline app/src/utils/formUtilities.ts:101]: Should there be a default? 'unknown user'?

**Review by stevecassidy — COMMENTED** [review]
Generally good with some suggestions for improvement and a few things to fix.  Ok if the suggestions become future tasks.

**Comment by stevecassidy** [comment]
I'm happy with this now. Very thorough work as usual!
